software/controller/lib/hal/adc.cpp

```diff
@@ -137,26 +261,25 @@ static constexpr int AdcConversionTime = [] {
   __builtin_unreachable();
 }();
 
-// Calculate how long our history buffer needs to be based on the above.
-static constexpr uint32_t AdcSampleHistory = static_cast<uint32_t>(
-    SampleHistoryTimeSec * CPUFrequencyHz / AdcConversionTime / OversampleCount / AdcChannels);
-
-// This scaler converts the sum of the A/D readings (a total of
-// AdcSampleHistory) into a voltage.  The A/D is scaled so a value of 0
-// corresponds to 0 volts, and MaxAdcReading corresponds to 3.3V
-static constexpr float AdcScaler = 3.3f / (MaxAdcReading * AdcSampleHistory);
+/// \TODO: have caller provide mappings in a different layer
+bool ADC::initialize(const uint32_t cpu_frequency_hz) {
+  adc_sample_history_ =
+      static_cast<uint32_t>(SampleHistoryTimeSec * static_cast<float>(cpu_frequency_hz) /
+                            AdcConversionTime / OversampleCount / AdcChannels);
+
+  // This scaler converts the sum of the A/D readings (a total of
+  // adc_sample_history_) into a voltage.  The A/D is scaled so a value of 0
+  // corresponds to 0 volts, and MaxAdcReading corresponds to 3.3V
+  adc_scaler_ = 3.3f / static_cast<float>(MaxAdcReading * adc_sample_history_);
+
+  // NOTE - we need the sample history to be small for two reasons:
+  // - We sum to a 32-bit floating point number and will lose precision if we add in too many
+  // samples
+  // - We want the A/D reading to be fast, so summing up a really large array might be too slow.
+  //
+  // If you get hit with this assertion you may need to rethink the way this function works.
+  if (adc_sample_history_ > 100) return false;
```

Comment:
So I have made it non-discard and added some comments around it now. Also defined the hard 100 limit as a static const. The HAL initializzer still does nothing withe the return value but at least there are bitchy TODOs around.

I am actually not sure on the correct course of action. The temptation is to template out these size limits, but then it would have to be explicitly instantiated in the caller and thus would be harder to make it a singleton. We will come back to this module's fate in a separate PR.